### ISSUE TITLE
Performance concerns: Excessive re-rendering of <Form> and <Field> components on unrelated redux store changes

### ISSUE THREAD

**Issue by JZBo** [rofrischmann/react-controlled-form#39]
Greetings,

While using the latest version of RCF, I noticed a weird behaviour of the ```<Form>``` component, which first I will summarize, then try to explain in more detail.

### Summary of the problem:
```<Form>``` performs an re-render from its own inner "render prop" on EVERY redux action dispatched that alters the store, regardless of where said alteration takes place. This causes a lot of unnecessary re-renders on unrelated user interaction, and I am concerned they might negatively affect performance.

### Details:
During a checkup on performance in a page when there were two different forms being shown, in completely different parts of the tree, I saw that the Forms would call their own **render props**, thus triggering an internal re-render, even on input on the OTHER Form. Then I noticed that happened also when other Redux actions altering the store were dispatched in the same page.
At first I thought it was related to the connected Container component ancestor (grandfather of both forms) in the page, but I verified it was not the case, even using advanced react-redux connect options like _arestateEqual_: the re-renders originated inside the Form (which is a connected component of his own) render props.

Considering that I first noticed this behaviour in what looked a somewhat specific edge case (two different forms on the same page), I made a temporary experiment in a simpler setting.
The test page tree has two branches: on one there is a children component, which has inside its tree a simple Form (I even removed any passed down props to the Form like other action dispatchers). 
On the other branch, there is a Button dispatching a very simple action to Redux (it simply toggles a boolean in an unrelated part of the store, simplest thing that came into my mind). 

Some code snippets of the 'simpler' test case to explain myself better:

```javascript
// Main Page
const MainArea = (props) => {
	const { className, manChanged, submitAC, children, ...others } = props;
	console.warn('------- MAIN AREA IS RE-RENDERING ----------');
	return (
		<div className={className}>
			<div><i> World Main Area </i></div>
			<Fragment forRoute={wk.NEW}>
				<NewForm />
			</Fragment>
			<GenericButton onClick={() => {
				console.log('CLICK CLICK');
                                // ManChanged is a dispatcher to redux;
				manChanged();
			}}>
				Test RE RENDERING BUTTON.
			</GenericButton>
		</div>
	);
};
```
```javascript
// Written as a Component since I used it to check inside its ShouldComponentUpdate
class NewForm extends React.Component {	
// ....
	render () {
		console.warn('NF RENDERING');
		return (
			<div className={this.props.className}>
				<Form
					formId="NewForm"
					initialFields={defaultFormFields()}
					render={
						(formRP) => {
							const {
								state, formId, data, updateField, updateState, reset, validate,
							} = formRP;
							console.warn('NF RENDER PROPS RENDERING');
							return (
								<NewFormInner
									formState={state}
									data={data}
									reset={reset}
									validate={validate}
									formId={formId}
									updateField={updateField}
									updateState={updateState}
								/>
							);
						}
					}
				/>
			</div>
		);
	}
}
```
(NewFormInner is a generic form made with React-Controlled-Form)

Now, when I click on said Button, the action triggers a re-render for the Form render props (and no re-render for any other connected components, if any are added, as it should be).

I'll add a simple screen of the console to show what happens.
![bugreport rcf 2018-09-05](https://user-images.githubusercontent.com/36004602/45087699-19c49080-b107-11e8-991e-42866a9dcc09.png)

At first you can see the form rendering normally after the app reaches the test location.
Then I perform some clicks on the Button, which should NOT trigger any kind of re-render by the Form, however, it does so every time.

**Comment by JZBo** [comment]
I can also suggest a temporary solution to other users of this fantastic package (which, for what it's worth, I really think is one of the best of all the redux aware form implementations out there).

By adding a shouldComponentUpdate check to the inner Form component (that is, to the Component being passed to the Form render props), it is possible to stop this excessive re-rendering.

For instance, in the example shown above, this solves the problem of excessive re-rendering:
```javascript
import dei from 'deep-equal-ident';
//.....
class NewFormInner extends React.Component {
	constructor (props) {
           // React doing its things              
	}
	
        // Intentionally verbose for the sake of example
	shouldComponentUpdate (nextProps, nextState) {
		const isStateEqual = dei(nextState, this.state);
		const isFormStateEqual = dei(nextProps.formState, this.props.formState);
		const isFormDataEqual = dei(nextProps.data, this.props.data);

		const shouldUpdate = !(isStateEqual && isFormDataEqual && isFormStateEqual);
		console.log(`-------> New Form Inner, SCU ${shouldUpdate}: stateEq ${isStateEqual}, formStateEq ${isFormStateEqual}. formDataEq ${isFormDataEqual}`);
		return shouldUpdate;
	}
//....
}
```

Same example as above:
![bugreport rcf 2018-09-05 - 02](https://user-images.githubusercontent.com/36004602/45088379-18946300-b109-11e8-9d9c-35cad810b200.png)

As it might be seen, this 'temporary fix' does not block the ```<Form>``` component calling for its own re-render, but at least blocks any useless-rendering of the component being returned by the Form render props.

(Of course, there is the usual trade-off of being forced to check shouldComponentUpdate, which means that any equality checks done in there will still affect performance).

**Comment by JZBo** [comment]
Hopefully I managed to explain myself, the problem and to report correctly the results of my investigations.

If there's any clarification, please ping me and I'll try to help debugging and solving it to the best of my abilities.

**Comment by robinweser** [comment]
Thanks for that huge detailed issue! I am right now refactoring and experimenting with a more simple and performant solution using the exact same API (so the update is going to be a drop-in replacement) that doesn't use Redux at all (but maybe is able to connect to Redux anyways).
Will be React > 16.3 only though...

Apart from that, I guess we could do some quick fixes to at least only update the components if the actual form is altered. Seems like an unintended bug to update on every Redux update.

**Comment by JZBo** [comment]
> Thanks for that huge detailed issue! I am right now refactoring and experimenting with a more simple and performant solution using the exact same API (so the update is going to be a drop-in replacement)    

That's good to know, and thanks again for your time in developing this very well thought out module. 
I'm especially glad to hear that it will keep the same API and will not require an overhaul on part of the users. It's really important for adopters to be able to rely on a somewhat stable API surface.

> that doesn't use Redux at all (but maybe is able to connect to Redux anyways).      

I'm slightly curious: if you don't mind me asking, what are the reasons to leave out Redux and what do you plan to use? 
It was quite useful having the form state in a predictable position in the Redux store: for instance, you could access it or compute some derived data from it anywhere in your app (e.g: an informative widget or tooltip shown on the same page, but located in a different, and mutable, part of the tree, like a dynamic toolbar).
I realize it would still be possible to do so 'by hand', by dispatching the relevant updates to a target slice of the store if and when they are needed, but that seems a bit redundant if the form state is still stored somewhere.

> Apart from that, I guess we could do some quick fixes to at least only update the components if the actual form is altered. Seems like an unintended bug to update on every Redux update.     

Some quick fixes would certainly be most welcome, as for sure they would be more effective on blocking re-render via shouldComponentUpdate in the render props components.
An hypothesis that came into my mind now: could it be it's some of the functions passed to the render props (like the ```validate``` one) that are found to be 'different' at every store update?

**Comment by JZBo** [comment]
Hi again. I wanted to add that I have had the occasion to check even more in depth, and it seems that this problem of excessive, useless re-rendering for completely unrelated actions being dispatched to any slice the store and processed by any unrelated reducer is not only at the ```Form``` level, but also at the ```Field``` one. 
It is possible to replicate the above-reported issue pretty much on the same terms also for Field, which is also a connected component like Form is. 
Thinking about it, it makes sense since the bug is also in the other component. Thus, it seems likely that the problem is in the way mappings are specified and computed for Form and Field.

However, this makes the issue **even more serious**, because even the temporary fix mentioned above on the Form then becomes quite useless to solve or dampen the sizeable performance issues caused by this constant unrelated re-rendering on every field for every redux driven interaction anywhere in the same page..... unless all Fields renders are manually wrapped in components implementing a custom sCU, which sounds like a nightmare of boilerplate and redundancy.

Is there any news on this? It has the potential to be a crippling problem for any extended use of this library.

**Comment by JZBo** [comment]
I have opened a pull request that mitigates a lot the problem by using a shouldComponentUpdate check on the Field component. Unfortunately I couldn't do the same for the Form component, as the initialization seems a bit tricky and didn't allow me to use the same technique  for attempting to resolve this issue.

Obviously, feel free to refuse it, I had mostly done it just to help you see where the problem is.

I also apologize if I messed up something, I'm not very used to the github site interface.

**Comment by JZBo** [comment]
Hello again,

I have had the occasion to investigate further on this issue, and I have precisely pinpointed the source of the problem. Took me a while to do so, sorry!
```Field``` and ```Form``` components are re-rendering too much (once on every redux interaction, whether related to them or not) because of the way the ```mapDispatchToProps```  for each of them is used with their ```connect``` to redux. 
Specifically, it is used with an arity of 2, and the presence of the ownProps parameter of the function means that the functions passed as props are not cached, thus being re-bound every time.
See, for instance, [these](https://stackoverflow.com/questions/46911024/how-do-i-avoid-re-rendering-a-connected-react-purecomponent-due-to-mapdispatchto/46918014#46918014) [two](https://github.com/reduxjs/react-redux/issues/157) comments on the topic. See also the official docs for [mapStateToProps](https://github.com/reduxjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options)

That was the cause of all the uncontrolled and performance-killing re-rendering happening that I reported in the first posts.

I have closed my previous pull request and put forward another with a more satisfying temporary solution, at least for the ```Field``` component, where I'm using ```shouldComponentUpdate``` to check for everything BUT the dispatch bound actions.

However, a much better long term solution, at least until any new version that makes redux optional, would be to avoid using ownProps at all. While it is true that the functions being bound to the dispatch are reliant on certain props of the field, it's also true that simpler workarounds are possible, by creating the bound functions to be passed down to the renderProp inside the Field component instance, which would be a faster and far more efficient solution.

Hopefully I explained with sufficient clarity what I meant, but if that was not the case, I'm all ears and all for helping. I really love this library :+1:
